### PR TITLE
Improve filters: references + multiple values + more predicate

### LINKED ISSUE #200: Update process compilation to add new filter functionalities
Update process compilation to add new filter functionalities.
Linked with issue https://github.com/mesg-foundation/engine/issues/1734 and PRs:
- https://github.com/mesg-foundation/engine/pull/1737
- https://github.com/mesg-foundation/engine/pull/1738
- https://github.com/mesg-foundation/engine/pull/1739

Work to do

- [x] support of basic types for filters (string, bool, number) (https://github.com/mesg-foundation/js-sdk/pull/203)
- [x] new definition for the process yml to handle references (https://github.com/mesg-foundation/js-sdk/pull/205)
- [ ] support of complex types for filters (array, object) (https://github.com/mesg-foundation/js-sdk/issues/209)
- [ ] support any ancestor for references (stepKey / nodeKey) (https://github.com/mesg-foundation/js-sdk/issues/210)

### PR DESCRIPTION
**Pull request by antho1404**
dependency: 
- https://github.com/mesg-foundation/engine/pull/1739
- https://github.com/mesg-foundation/engine/pull/1754
- https://github.com/mesg-foundation/js-sdk/pull/207

related to:

- https://github.com/mesg-foundation/engine/issues/1734
- https://github.com/mesg-foundation/js-sdk/issues/200

## Filter any value

You can now create a filter with conditions that are not only based on string value but also boolean and number.

Now the following filter in a process will be working

```yml
- type: filter
  conditions:
    foo: true
    bar: 42
    hello: "world"
```

We might add the support of array/object in another PR

## Reference

Add the possibility to define an advanced condition in the filter.

Current filter (still valid and will be automatically converted in the new advanced filter):
```yaml
- type: filter
  conditions:
     foo: bar
```

This is equivalent to
```yaml
- type: filter
  conditions: 
    - key: foo
      predicate: EQ
      value: bar
```

This format lets you create much-advanced filters.
You can now add references (https://docs.mesg.com/guide/process/steps/task.html#input) as a key to reference a nested data, you can also add any predicate from the following list (https://github.com/mesg-foundation/js-sdk/blob/c44815aa1aa3a81a60803b7b8d7f2f35bf712416/packages/api/src/process-lcd.ts#L112-L119)
and a value that can be anything based on https://github.com/mesg-foundation/js-sdk/pull/203

### What's in this PR:
- [x] update LCD types
- [x] update compiler to use new types
- [x] update the process definition to accept any value for the filter
- [x] update compiler to add any value for the filter
- [x] support reference in the compilation (need to update the definition)

### To test

You can use the following process:
```yml
name: Test
steps:
  - type: trigger
    instance:
      src: https://github.com/mesg-foundation/service-webhook
    eventKey: request
  - type: filter
    conditions:
      - key: data.foo[0].x
        predicate: EQ
        value: 2
  - type: task
    instance:
      src: https://github.com/mesg-foundation/service-js-function
    taskKey: execute
    inputs:
      code: "module.export = function(value) { return true }"
      inputs:
        foo: 42
```

and run 
```bash 
mesg-cli-dev process:dev ~/Desktop/simple-process.yml --image mesg/engine:local
```

then test with 
```bash
curl -XPOST http://localhost:3005/webhook --data '{"foo": [{"x": 2}]}' -H "Content-Type: application/json"
```

of course, you can play with different conditions and different data

### CODE REVIEW

File: packages/api/CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@
 - [#188](https://github.com/mesg-foundation/js-sdk/pull/188) Add mnemonic generation to account
 - [#199](https://github.com/mesg-foundation/js-sdk/pull/199) Add helper to get a hash of a resource created from a tx result
 - [#199](https://github.com/mesg-foundation/js-sdk/pull/199) Add create/delete/hash messages for runners
+- [#203](https://github.com/mesg-foundation/js-sdk/pull/203) Update LCD types and add process filters value and reference
```
Comment (antho1404) [inline packages/api/CHANGELOG.md]:
```suggestion
- [#205](https://github.com/mesg-foundation/js-sdk/pull/205) Update LCD types and add process filters value and reference
```

File: packages/compiler/CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,19 @@
 - [#188](https://github.com/mesg-foundation/js-sdk/pull/188) Compile services in a LCD format
 
 #### Improvements
+
+- [#203](https://github.com/mesg-foundation/js-sdk/pull/203) Add support of any value for filters value (string, boolean or number)
+- [#](https://github.com/mesg-foundation/js-sdk/pull/) Add advance filter conditions
```
Comment (antho1404) [inline packages/compiler/CHANGELOG.md]:
```suggestion
- [#205](https://github.com/mesg-foundation/js-sdk/pull/205) Add support of any value for filters value (string, boolean or number)
- [#205](https://github.com/mesg-foundation/js-sdk/pull/205) Add advance filter conditions
```

**Review by NicolasMahe — APPROVED** [review]
well done :)

**Comment by antho1404** [comment]
@blackinno any feedback on how to improve this and if this will solve most of the issues you had? (eg: https://github.com/mesg-foundation/engine/issues/1582)

**Comment by blackinno** [comment]
> @blackinno any feedback on how to improve this and if this will solve most of the issues you had? (eg: [mesg-foundation/engine#1582](https://github.com/mesg-foundation/engine/issues/1582))

Nice to have and `filter` like this will resolve my issues

**Pull request merged**